### PR TITLE
Fix NaN gradients in nn.MultiheadAttention for fully-masked rows (#41…

### LINKED ISSUE #41508: nn.MultiheadAttention causes gradients to become NaN under some use cases
## 🐛 Bug

Using key_padding_mask and attn_mask with nn.MultiheadAttention causes gradients to become NaN under some use cases.

## To Reproduce

Backwards pass through nn.MultiheadAttention layer where the forward pass used:
1. attn_mask limiting context in both directions (e.g. bucketed attention)
2. key_padding_mask where there is padding for at least one sequence (and there is also at least one valid entry for every sequence, as expected)
3. The timesteps that were masked are not used to calculate the loss

import torch
torch.manual_seed(0)
'''Create attention layer'''
attn = torch.nn.MultiheadAttention(embed_dim=1, num_heads=1)
'''Create dummy input'''
x = torch.rand(3, 2, 1)
'''Padding mask, second sequence can only see first embedding'''
key_padding_mask = torch.as_tensor([[False, False, False], [False, True, True]], dtype=torch.bool)
'''Attention mask, bucketing attention to current and previous time steps'''
attn_mask = torch.as_tensor([[0., float('-inf'), float('-inf')], [0., 0., float('-inf')], [float('-inf'), 0., 0.]])
'''Generate attention embedding'''
output, scores = attn(x, x, x, key_padding_mask=key_padding_mask, attn_mask=attn_mask)
print("scores")
print(scores)
'''Create a dummy loss, only use the first embedding which is defined for all sequences'''
loss = output[0, :].sum()
print("loss")
print(loss)
'''Backwards pass and gradients'''
loss.backward()
print("grads")
for n, p in attn.named_parameters():
    print(n, p.grad)

> scores
> tensor([[[1.0000, 0.0000, 0.0000],
>          [0.4468, 0.5532, 0.0000],
>          [0.0000, 0.5379, 0.4621]],
>         [[1.0000, 0.0000, 0.0000],
>         [1.0000, 0.0000, 0.0000],
>          [   nan,    nan,    nan]]], grad_fn=<DivBackward0>)
> loss
> tensor(0.0040, grad_fn=<SumBackward0>)
> grads
> in_proj_weight tensor([[nan],
>         [nan],
>         [nan]])
> in_proj_bias tensor([nan, nan, nan])
> out_proj.weight tensor([[nan]])
> out_proj.bias tensor([2.])

## Expected behavior

Gradients should not be NaN

## Environment

PyTorch version: 1.5.1
Is debug build: No
CUDA used to build PyTorch: None

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.5.1
[conda] blas                      1.0                         mkl
[conda] cpuonly                   1.0                           0    pytorch
[conda] mkl                       2020.1                      217
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.1.0            py37h23d657b_0
[conda] mkl_random                1.1.1            py37h0573a6f_0
[conda] numpy                     1.18.5           py37ha1c710e_0
[conda] numpy-base                1.18.5           py37hde5b4d6_0
[conda] pytorch                   1.5.1               py3.7_cpu_0  [cpuonly]  pytorch

Also fails when using GPU.

### PR DESCRIPTION
**Pull request by Hemanth-1354**
---Fixes #41508

Root cause

nn.MultiheadAttention produces NaN gradients when a key_padding_mask (or attn_mask) fully masks out every key for a given query row. Attention scores for that row become all -inf, and softmax(-inf, ..., -inf) is 0/0 NaN. That NaN propagates through backward, corrupting gradients for the whole batch even though only one row was invalid.

SDPA fast path (need_weights=False) already avoids this via torch._safe_softmax in aten/src/ATen/native/transformers/attention.cpp, zeroing fully-masked rows. Eager path in multi_head_attention_forward (need_weights=True) still called plain softmax, missing that protection.

Fix

- torch/nn/functional.py: switch eager path's softmax to torch._safe_softmax.
- test/test_transformers.py: added test_multiheadattention_fully_masked_row_no_nan_grad, parametrized over need_weights/batch_first, checks output+gradient NaN-free.
- test/nn/test_multihead_attention.py: existing numpy reference softmax computed plain 0/0 for fully-masked rows, matching old buggy NaN. Now torch returns zero row, so reference updated same way, mirroring aten._safe_softmax decomposition.

Test plan

- python test/test_transformers.py -v -k test_multiheadattention_fully_masked_row_no_nan_grad
- python test/nn/test_multihead_attention.py -v -k average_attn_weights
---

### CODE REVIEW

**Comment by ezyang** [comment]
@claude does this seem right

**Comment by drisspg** [comment]
@claude this code has been frozen. There are a number of weird branches that you can take depdneing on coditions for nn.MHA. Some got sdpa (math) or sdpa(fused) which do have the rows -> 0 output mapping. Others go to exsting better transformer paths that dont. Can you draw out the full map for me explain in great detail and what this PR does. I historically havent landed things since its frozen but I might reconsider if you paint the picture for me

**Review by drisspg — APPROVED** [review]
Lets party

**Comment by drisspg** [comment]
failures look real

**Comment by Hemanth-1354** [comment]
@drisspg  pushed a fix for the autocast dtype mismatch, the `_safe_softmax` decomposition (used under fake-tensor/proxy tracing) wasn't pinning output dtype, so autocast's fp32 upcast policy was leaking through and causing the float16 vs float32 mismatch in `test_fake_autocast_nn_functional_multi_head_attention_forward`. Should be fixed now.
@drisspg 
Could you approve the pending CI workflows on this PR?

**Comment by Hemanth-1354** [comment]
Hi @drisspg! Just following up on this PR. If there's anything you'd like me to change or improve, I'd be happy to make the updates. Thanks for taking the time to review it!
